[Merged by Bors] - p2p periodically check peers from addressBook for dead|alive

abergasov opened the pull request:
## Motivation
Closes #3132

## Changes
added method which in ticker take random X addresses and sen request for it

## Test Plan
unit tests

## DevOps Notes
<!-- Please uncheck these items as applicable to make DevOps aware of changes that may affect releases -->
- [x] This PR does not require configuration changes (e.g., environment variables, GitHub secrets, VM resources)
- [x] This PR does not affect public APIs
- [x] This PR does not rely on a new version of external services (PoET, elasticsearch, etc.)
- [x] This PR does not make changes to log messages (which monitoring infrastructure may rely on)

abergasov committed:
check book peers

abergasov committed:
test book check

abergasov committed:
linter fix

abergasov committed:
tests add

abergasov committed:
tests add

abergasov committed:
address book sync from peerBook

dshulyak commented on p2p/nodechecker_test.go:
```diff
@@ -0,0 +1,212 @@
+			block:    []int{},
+		},
+	}
+	for _, testCase := range table {
```
testCase needs to be moved out of loop if t.Parallel is used, otherwise it will run the same test case but print it with a different name.

```
=== RUN   TestHost_LocalAddressChange
=== RUN   TestHost_LocalAddressChange/change_local_address_of_node_and_all_not_acceptable
=== PAUSE TestHost_LocalAddressChange/change_local_address_of_node_and_all_not_acceptable
=== RUN   TestHost_LocalAddressChange/change_local_address_of_node_and_old_not_acceptable
=== PAUSE TestHost_LocalAddressChange/change_local_address_of_node_and_old_not_acceptable
=== RUN   TestHost_LocalAddressChange/change_local_address_of_node_and_new_not_acceptable
=== PAUSE TestHost_LocalAddressChange/change_local_address_of_node_and_new_not_acceptable
=== RUN   TestHost_LocalAddressChange/change_local_address_of_node_and_all_are_acceptable
=== PAUSE TestHost_LocalAddressChange/change_local_address_of_node_and_all_are_acceptable
=== CONT  TestHost_LocalAddressChange/change_local_address_of_node_and_all_not_acceptable
ACTUAL TEST CASE: change local address of node and all are acceptable
=== CONT  TestHost_LocalAddressChange/change_local_address_of_node_and_new_not_acceptable
ACTUAL TEST CASE: change local address of node and all are acceptable
=== CONT  TestHost_LocalAddressChange/change_local_address_of_node_and_old_not_acceptable
ACTUAL TEST CASE: change local address of node and all are acceptable
=== CONT  TestHost_LocalAddressChange/change_local_address_of_node_and_all_are_acceptable
ACTUAL TEST CASE: change local address of node and all are acceptable
```
also please run a test with -race

abergasov replied:
race generate my helper struct with blackList addresses, when access without mutex

dshulyak commented on p2p/peerexchange/bookchecker.go:
```diff
@@ -0,0 +1,89 @@
+	qCtx, cancel := context.WithTimeout(context.Background(), checkTimeout)
+	defer cancel()
+	if _, err := d.crawl.query(qCtx, peers); err != nil {
+		d.logger.Error("failed to query bootstrap node: %s", err)
```
error message?

dshulyak commented on p2p/peerexchange/bookchecker.go:
```diff
@@ -0,0 +1,89 @@
+	}
+}
+
+func (d *Discovery) CheckPeers() {
```
new context should wrap context that is used by errogroup

dshulyak commented on p2p/peerexchange/bookchecker.go:
```diff
@@ -0,0 +1,89 @@
+		select {
+		case <-ticker.C:
+			d.logger.Debug("checking book")
+			d.CheckPeers()
```
ctx to CheckPeers

dshulyak commented on p2p/peerexchange/bookchecker.go:
```diff
@@ -0,0 +1,89 @@
+const (
+	checkInterval = 3 * time.Minute  // Interval to check for dead|alive peers in the book.
+	checkTimeout  = 30 * time.Second // Timeout to check for dead|alive peers in the book.
+	peersNumber   = 10               // Number of peers to check for dead|alive peers in the book.
```
can be it be a part of config?

dshulyak reviewed (changes requested):
1. we should protect bootnodes from being removed from address book, even if they are temporarily unavailable
2. if connection with a peer exists peer should not be queried
3. if peer was queried not long time (maybe 30m-60m?) ago we shouldn't query it again
4. each query will create a connection, i think checker should disconnect a peer if connection was specifically created for a short check

abergasov committed:
close connection after check
use condition for peers ping select
config as dependency
parallel tests fix
node condition select tests

abergasov committed:
Merge branch 'develop' into 3132-p2p-stale-records-in-the-address-book

abergasov committed:
linter fix

dshulyak commented on p2p/nodechecker_test.go:
```diff
@@ -0,0 +1,241 @@
+
+func discoveryBootstrap(disc *peerexchange.Discovery) {
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
```
why it needs defer cancel() if function exits only if Bootstrap is canceled?

abergasov replied:
removed this

dshulyak commented on p2p/peerexchange/bookchecker.go:
```diff
@@ -0,0 +1,133 @@
+		return
+	}
+	// check peers are updated in host peerBook.
+	for peerID, addresses := range d.peersToMap(peers) {
```
i dont understand why addresses needs to be an array. we get N random peers on L35. is it possible for that result to contain duplicates?

abergasov replied:
can be situation, when some nodeA update local address. After updating - node will be trigger notify event and peerBook of nodeB add new address and will contain multiply addresses for nodeA. that's why peerBook contains multiply addresses per node

while our addreesBook store address in `map[peerID]*addrInfo`. in method `GetRandomPeers` L105 i check, does peerBook have extra address for nodeA. if so - add additional address for check.

than check all addresses of node. if one of them is not available - remove. if only one - update to alive address

dshulyak commented on p2p/peerexchange/bookchecker.go:
```diff
@@ -0,0 +1,133 @@
+			continue
+		}
+		// update node address in book.
+		d.book.AddAddress(newAddr, addresses[1])
```
why is it addresses[1]? should it be addresses[0]? but maybe we can get rid from duplicates in our peer book instead

abergasov replied:
yes, forgot rollback after tests. `addresses[0]` should be here

in case of multiply addresses valid after check - just take first. in our book will be only one address

abergasov committed:
review fix

countvonzero commented on .github/workflows/ci.yml:
```diff
@@ -65,7 +65,6 @@ jobs:
           make
           make tidy
           make test-fmt
-          make lint
```
we are removing lint check? why?

abergasov replied:
We use Golint in this command. This tool is deprecated and frozen. Also this tool have not annotations
We also using golangci-lint instead as separate step `make golangci-lint-github-action`

It just cleanup)

abergasov commented:
bors try

abergasov commented:
bots try

abergasov commented:
bors try

abergasov commented:
bors merge
